src/render_helpers/mod.rs

```diff
@@ -101,14 +85,69 @@ pub fn render_to_dmabuf(
     size: Size<i32, Physical>,
     scale: Scale<f64>,
     elements: impl Iterator<Item = impl RenderElement<GlesRenderer>>,
-) -> anyhow::Result<()> {
+) -> anyhow::Result<SyncPoint> {
     let _span = tracy_client::span!();
+    renderer.bind(dmabuf).context("error binding texture")?;
+    render_elements(renderer, Transform::Normal, scale, size, elements)
+}
 
+/// # Safety
+///
+/// The caller must ensure that the provided WlBuffer is the correct size and format.
+pub unsafe fn render_to_shm(
+    renderer: &mut GlesRenderer,
+    buffer: &WlBuffer,
+    size: Size<i32, Physical>,
+    scale: Scale<f64>,
+    transform: Transform,
+    elements: impl Iterator<Item = impl RenderElement<GlesRenderer>>,
+) -> anyhow::Result<SyncPoint> {
+    let buffer_size = smithay::backend::renderer::buffer_dimensions(buffer).unwrap();
+    let offscreen_buffer: GlesRenderbuffer = renderer
+        .create_buffer(Fourcc::Abgr8888, buffer_size)
+        .context("error creating renderbuffer")?;
+    renderer
+        .bind(offscreen_buffer)
+        .context("error binding renderbuffer")?;
+
+    let sync_point = render_elements(renderer, transform, scale, size, elements)?;
+
+    shm::with_buffer_contents_mut(buffer, |shm_buffer, shm_len, buffer_data| {
+        debug_assert!(
+            // The buffer prefers pixels in little endian ...
+            buffer_data.format == wl_shm::Format::Argb8888
+                && buffer_data.stride == size.w * 4
+                && buffer_data.height == size.h
+                && shm_len as i32 == buffer_data.stride * buffer_data.height
+        );
+        renderer.with_context(|gl| unsafe {
+            gl.ReadPixels(
+                0,
+                0,
+                size.w,
+                size.h,
+                // ... but OpenGL prefers them in big endian.
+                gles::ffi::BGRA_EXT,
+                gles::ffi::UNSIGNED_BYTE,
+                shm_buffer.cast(),
+            )
+        })?;
+
+        Ok(sync_point)
```

Comment:
`glReadPixels` is synchronous, it won't return until the operation completes. So the sync point is useless here (in this case, `glReadPixels` itself will wait for that sync point before copying data to the main memory).

Reply:
Hm. so it does seem. is there an equivalent asychronous call? Should the `glReadPixels` be moved into *after* the `SyncPoint`? Does it really matter?

Reply:
The correct thing to do here is to simply do nothing with the sync point and return None. The sync point should be used only for the dmabuf buffer type.